client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java

```diff
@@ -196,4 +229,19 @@ public static Position fromBytes(ByteBuffer buff) {
         return SERIALIZER.deserialize(new ByteArraySegment(buff));
     }
 
+    private synchronized void applySegmentOffsetUpdatesIfNeeded() {
+        // No updates, so nothing to do.
+        if (this.updatesToSegmentOffsets == null || this.updatesToSegmentOffsets.isEmpty()) {
+            return;
+        }
+        // We are going to modify ownedSegments to match the segment offsets at the time this event was read, so copying
+        // the map of ownedSegments to do not impact to the map referenced by other PositionImpl objects.
+        this.ownedSegments = new HashMap<>(ownedSegments);
+        // Apply all the Segment offset updates up to the point in which this event was read.
+        for (int i = 0; i < this.version; i++) {
+            this.ownedSegments.put(this.updatesToSegmentOffsets.get(i).getKey(), this.updatesToSegmentOffsets.get(i).getValue());
```

Comment:
@shrids 's intuition was correct. This call to `updatesToSegmentOffsets.get(i)` is not thread safe. The list may not be synchronized in any way between the thread calling readNextEvent and the thread calling this method. As such this could endup reading garbage data.

Reply:
You are right. In the case of using an `ArrayList` in that way, an internal resize can be triggered by one thread adding new elements, which might lead another thread to read garbage when executing this method in `PositionImpl` that reads that list. A solution to this is to use an "structurally immutable" list, wrapped by a `List` for convenience. This is currently achieved by using `ArrayList.asList()` for initializing `updatesToSegmentOffsets`. Any attempt to `add` or `delete` elements from the list will result in `UnsupportedOperationException`, while `set` and `get` methods are allowed (similar to an array).